docs/src/main/asciidoc/application-data-caching-guide.adoc

```diff
@@ -0,0 +1,474 @@
+If the `@CacheLoad`, `@CacheStore` or `@CacheInvalidate` annotations are used on a method with no parameters, a unique default
+cache key derived from the cache name will be generated and used.
+====
+
+=== @CacheInvalidateAll
+
+When a method annotated with `@CacheInvalidateAll` is invoked, Quarkus will remove all entries from the cache.
+
+[TIP]
+====
+Each of the `@CacheLoad`, `@CacheStore`, `@CacheInvalidate` and `@CacheInvalidateAll` annotations takes an optional `cacheName`
+parameter which can be used to specify the name of the cache. This is especially useful when the same cache must be accessed
+from several classes. If the cache name is not specified, the fully qualified name of the class where the annotated method is
+declared will be used as the default cache name.
+====
+
+=== @CacheKey
+
+When a method argument is annotated with `@CacheKey`, it is identified as a part of a cache key during an invocation of a
+method annotated with `@CacheLoad`, `@CacheStore` or `@CacheInvalidate`.
+
+This annotation is optional and should only be used when some of the method arguments are NOT part of the cache key.
+
+== Configuring the underlying caching provider
+
+For now, https://github.com/ben-manes/caffeine[Caffeine] is the only available local caching provider in Quarkus. We plan to
+add more providers soon.
+
+=== Caffeine configuration properties
+
+Caffeine is a high performance, near optimal caching library based on Java 8.
+
+Each of the Caffeine caches backing up the Quarkus application data caching extension can be configured using the following
+properties in the `application.properties` file.
+
+[TIP]
+====
+You need to replace `cache-name` in all of the following properties with the real name of the cache you want to configure.
+====
+
+`quarkus.cache.caffeine."cache-name".initial-capacity`::
+Minimum total size for the internal data structures. Providing a large enough estimate at construction time avoids the need for
+expensive resizing operations later, but setting this value unnecessarily high wastes memory.
+
+`quarkus.cache.caffeine."cache-name".maximum-size`::
+Maximum number of entries the cache may contain. Note that the cache may evict an entry before this limit is exceeded or
+temporarily exceed the threshold while evicting. As the cache size grows close to the maximum, the cache evicts entries that
+are less likely to be used again. For example, the cache may evict an entry because it hasn't been used recently or very often.
+
+`quarkus.cache.caffeine."cache-name".expire-after-write`::
+Specifies that each entry should be automatically removed from the cache once a fixed duration has elapsed after the entry's
+creation, or the most recent replacement of its value.
+
+`quarkus.cache.caffeine."cache-name".expire-after-access`::
+Specifies that each entry should be automatically removed from the cache once a fixed duration has elapsed after the entry's
+creation, the most recent replacement of its value, or its last read.
```

Comment:
It would be nice if we could have this generated: you might want to have one short sentence then the details so that it looks better in the generated doc.

Reply:
I'll work on this point during the week-end using @machi1990's #4050.